### PR TITLE
fixes #4959 allow asterix literals in variable syntax

### LINKED ISSUE #4959: Asterisk in default variable value leads to value interpreted as string and not interpolated
```allowedOrigin: ${opt:allowedOrigin, 'beta.blabla.net'}```
works fine
```allowedOrigin: ${opt:allowedOrigin, '*.blabla.net'}``` 
 doesnt work

in second case allowedOrigin will be interpreted as a string e.g. its value will be ${opt:allowedOrigin, '.blabla.net'} instead of *.blabla.net

### PR DESCRIPTION
**Pull request by beeceej**
This PR addresses issue #4959.

The previous regex used to look for candidate variables was not allowing expressions with `*`'s, this PR changes all instances of said regex 

`\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}`
 to 
`\${(([ ~:a-zA-Z0-9._'",\\-\\/\\(\\)]|\*(?=([ ~:a-zA-Z0-9._'",\\-\\/\\(\\)]+?)('|")))+?)}` which allows for `*` to be used as values in the variable syntax, only if they are surrounded by `'` or `"`

## How can we verify it:
1. Observe that a function deployed using *code from this patch* will have the correct STAGE_VAR (`*.mystage.com`)
2. Observe that a function deployed using *code deployed without this patch* will have the incorrect STAGE_VAR (`${opt:stage, '*.mystage.com'}`)

```yaml
service: test-sls-changes

custom:
  stage: ${opt:stage, '*.mystage.com'}

provider:
  name: aws
  runtime: nodejs6.10
  environment:
    STAGE_VAR: ${self:custom.stage}

functions:
  hello:
    handler: handler.hello
```

## Todos:
- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by HyperBrain** [comment]
@beeceej Thanks for the improvement. Hoever I'd like to verify roughly if there was a special reson for excluding the asterisk in other areas first.
@erikerikson Can you remember any reason why it was not included, or were there any cases were it explicitly had been omitted due to the behavior of the resolved vars (the only thing that comes into my mind right now is file refs)

**Comment by erikerikson** [comment]
I did not originally construct that and don't know off the top of my head any reason for the restriction.  @eahefnawy do you have a reason for not including the `*` character in the variable regex?

**Comment by erikerikson** [comment]
This said, there should be at least some consideration for any circumstance where the SLS framework is used automatically as part of a CI/CD process with input from the public (e.g. automated processing of PRs).  There are security risks associated.

**Comment by HyperBrain** [comment]
@erikerikson 
> There are security risks associated.

Good point. We should make sure that an asterisk is not passed down from the variables to any file lookup - i.e. especially for the file references. The question is now, how to add such checks and from an architecture perspective, _where_ to add them. In my opinion, the variable parser only can allow them, but they have to be dropped at the locations they are actually used (e.g. in the file reference resolver).

**Comment by beeceej** [comment]
Very good points,

Are we thinking the criteria for acceptance would be shoring up of the file resolver (if needed), and the appropriate tests to go along with it?

I'm happy to continue with the extra work (it'll be a good way to dig more into the codebase).

At any rate, I'll wait for further input.

**Comment by HyperBrain** [comment]
The idea with the lookahead sound promising. If we would let the regex only match in case the asterisk is used in a quoted string, the regex itself would implicitly disallow wrong uses.
I think the asterisk only makes sense in a quoted string. All other uses within a variable syntax should lead to errors - given that variable names themselves are also not allowed to contain it.

**Comment by beeceej** [comment]
I will revisit the regex being used for the interpolation to ONLY allow `*` if it is used in a quoted string

as such I'm moving this PR out of ready to review for further work/tests. Thanks for the feedback!

**Comment by beeceej** [comment]
@HyperBrain @erikerikson 
 I've revisited the proposed regex and am now using the look ahead suggested.

I could be wrong, but it looks like some of the tests are flaking out. I do not mind adding more tests around this If needed. Let me know what the next steps should be.

File: lib/classes/Variables.test.js [hunk 1]
```diff
@@ -156,6 +158,30 @@ describe('Variables', () => {
...
+
+        awsProvider.options[property.name] = '${self:foobar, *.a.thing}';
+        serverless.variables.populateService().should.be.fulfilled
+          .then(() => expect(property.getter(awsProvider)).to.be.eql('${self:foobar, *.a.thing}'));
+      });
+
+      it('should populate variables containing * if it is sourrounded by \'', () => {
+        const property = { name: 'region', getter: (provider) => provider.getRegion() };
+
+        awsProvider.options[property.name] = '${self:foobar, "*.a.thing"}';
```
Comment (erikerikson) [inline lib/classes/Variables.test.js]: Minor nit: you seem to have the single and double quote cases reversed in name or implementation.

File: lib/classes/Variables.test.js [hunk 2]
```diff
@@ -156,6 +158,30 @@ describe('Variables', () => {
             .then(() => expect(property.getter(awsProvider)).to.be.eql('default'));
         });
       });
+
+      it('should not populate variables containing *', () => {
+        const property = { name: 'region', getter: (provider) => provider.getRegion() };
+
+        awsProvider.options[property.name] = '${self:foobar, *.a.thing}';
+        serverless.variables.populateService().should.be.fulfilled
+          .then(() => expect(property.getter(awsProvider)).to.be.eql('${self:foobar, *.a.thing}'));
```
Comment (erikerikson) [inline lib/classes/Variables.test.js]: So you can disable variable resolution by accidentally typing in an asterisk?  That may be fine but could confuse without offering warning.

Comment (beeceej) [reply lib/classes/Variables.test.js]: Good point.

The existing behavior is that variable resolution will be disabled if an `*` is in the declaration, regardless of `"` `'`.

It probably does make sense to add some warning.

Are there any existing PR's which would serve as a good template for adding such a warning?

Comment (erikerikson) [reply lib/classes/Variables.test.js]: Sorry, context loss error on my part.  As existing behavior, I would not expect you to fix it.  Your changes reduce the surface space of the unexpected/odd/confusion behavior and that's an improvement over current state.

**Comment by pmuens** [comment]
Thanks for taking the time to work on this @beeceej 👍

I'm closing this PR (as discussed) for https://github.com/serverless/serverless/pull/5640 where @dschep continued the work...

**Pull request closed without merging**